clients/src/main/java/org/apache/kafka/common/telemetry/ClientTelemetrySender.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.telemetry;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.annotation.InterfaceStability;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.requests.AbstractResponse;
+import org.apache.kafka.common.requests.AbstractRequest.Builder;
+
+import java.util.Optional;
+
+/**
+ * A {@link MetricsReporter} may implement this interface to indicate support for sending client
+ * telemetry to the broker.
+ */
+@InterfaceStability.Evolving
+public interface ClientTelemetrySender extends AutoCloseable {
+
+    /**
+     * Return the next time when the telemetry API should be attempted (i.e., interval time has elapsed).
+     * <p>
+     * If another telemetry API is in-flight, then {@code timeoutMs} should be returned as the
+     * maximum wait time.
+     *
+     * @param timeoutMs The timeout for the inflight telemetry API call.
+     * @return remaining time in ms till the telemetry API be attempted again.
+     */
+    long timeToNextUpdate(long timeoutMs);
+
+    /**
+     * Return the telemetry request based on client state i.e. determine if
+     * {@link org.apache.kafka.common.requests.GetTelemetrySubscriptionsRequest} or
+     * {@link org.apache.kafka.common.requests.PushTelemetryRequest} be constructed.
+     *
+     * @return request for telemetry API call.
+     */
+    Optional<Builder<?>> createRequest();
+
+    /**
+     * Handle response for telemetry APIs
+     *
+     * @param response either {@link org.apache.kafka.common.requests.GetTelemetrySubscriptionsResponse} or
+     *                 {@link org.apache.kafka.common.requests.PushTelemetryResponse} telemetry API response.
+     */
+    void handleResponse(AbstractResponse response);
```

Comment:
Is there any reason not to have overloaded "handle" methods for the two response types we expect the implementation to handle?

Reply:
In more precise terms, I'm thinking this should be two methods:

```java
    void handleResponse(GetTelemetrySubscriptionsResponse response);
    void handleResponse(PushTelemetryResponse response);
```

Reply:
That's what I had earlier and then thought to determine based on response type. The methods were guarded with a check to send only GetTele.... and PushTele... response type in this method. But it might be worth to tightly bind, I ll change the code.

Reply:
I will correct it once my first PR for telemetry json contracts are merged as they have dependency.

Reply:
Done.